[CodeCompletion] Some refactorings surrounding solver-based completion

ahoppen opened the pull request:
Perform a couple of refactoring for solver-based completion, with the goal of sharing more code between the different completion kinds:
- Sink `TypeCheckCompletionCallbacks` from Sema to IDE
  - That’s really where it belongs, there’s no reason for it to be in Sema
- Unify logic to retrieve completion expr type for all completion callbacks
  - Previously, each completion kind had its own handling of `UnresolvedType` and some had fewer fallback mechanisms to retrieve the type, have one implementation that does everything
- Check whether surrounding context supports async in all solver-based completion kinds
  - Previously, only `ExprCompletion` checked whether the surrounding context is async, I don’t have any failing test cases for the other completion kinds but it definitely makes sense for them to share the same logic to determine whether the surrounding context is async

ahoppen committed:
[CodeCompletion] Move TypeCheckCompletionCallbacks to IDE

ahoppen committed:
[CodeCompletion] Unify logic to retrieve completion expr type for all completion callbacks

ahoppen committed:
[CodeCompletion] Check whether surrounding context supports async in all solver-based completion kinds

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

xedin commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -207,6 +161,52 @@ bool swift::ide::isCompletionDeclContextLocalContext(DeclContext *DC) {
+    //   func foo(fn: () async -> Void)
+    //   foo { <HERE> }
+    // In this case, the closure is wrapped with a 'FunctionConversionExpr'
+    // which has 'async' function type.
```
This assume a type-checked version of the AST, is that always the case?

ahoppen replied:
I believe so. The call side inside `CompletionLookup` is intended for the legacy code completion, which works on a type-checked AST and https://github.com/apple/swift/pull/41885/files/03d819f442f05b58f6a2b1816b2e280292db5329#diff-3088e637c4f06d0fc6cb38370068cd6fa6fe8dcaa10ec35c9086602c59eaba17R149 works on the already type-checked part of the AST to determine async-ness after failing to determine async-ness from the solution.

xedin commented on include/swift/IDE/TypeCheckCompletionCallback.h:
```diff
@@ -0,0 +1,84 @@
+  /// Called for each solution produced while type-checking an expression
+  /// that the code completion expression participates in.
+  virtual void sawSolution(const constraints::Solution &solution) {
+    GotCallback = true;
```
This looks suspicious - there is no way to ensure that each override calls `TypeCheckCompletionCallback::sawSolution(solution)` internally. I see that they currently do but it only takes one to go unnoticed... I'd like to suggest `virtual` to be removed from this method, and instead be used on a `protected` `sawSolutionImpl(const Solution &)` which is going to be called from this one.

ahoppen replied:
Sounds like a good idea. I’ll do that.

ahoppen replied:
Implemented in https://github.com/apple/swift/pull/41885/commits/3597652960253685386103751dad1837fe410213

ahoppen committed:
[CodeCompletion] Make sawSolution non-final and override sawSolutionImpl from subclasses

This eliminates a source of bugs if subclasses of `TypeCheckCompletionCallback` forget to call the superclass’s implementation of `sawSolution` from their overridden method.

ahoppen commented:
@swift-ci Please smoke test

xedin commented on include/swift/IDE/ArgumentCompletion.h:
```diff
@@ -63,13 +63,13 @@ class ArgumentTypeCheckCompletionCallback : public TypeCheckCompletionCallback {
                          SmallVectorImpl<PossibleParamInfo> &Params,
                          SmallVectorImpl<Type> &Types);
 
+  void sawSolutionImpl(const constraints::Solution &solution) override;
```
Nit: mark it as pure via ` = 0;`

xedin replied:
Oops, wrong place :)
